ewatercycle/analysis/__init__.py

```diff
@@ -0,0 +1,136 @@
+import pandas as pd
+import os
+from typing import Union
+from hydrostats import metrics
+import matplotlib.pyplot as plt
+from matplotlib.dates import DateFormatter
+
+
+def hydrograph(
+    discharge: pd.DataFrame,
+    *,
+    reference: str,
+    precipitation: pd.DataFrame = None,
+    fname: Union[os.PathLike, str] = None,
+    dpi: int = None,
+    title: str = 'Hydrograph',
+    discharge_units: str = 'm$^3$ s$^{-1}$',
+    precipitation_units: str = 'mm day$^{-1}$',
+) -> plt.Axes:
+    """Plot a hydrograph.
+
+    This utility function makes it convenient to create a hydrograph from
+    a set of discharge data from a `pandas.DataFrame`. A column must be marked
+    as the reference, so that the agreement metrics can be calculated.
+
+    Optionally, the corresponding precipitation data can be plotted for
+    comparison.
+
+    Parameters
+    ----------
+    discharge : pd.DataFrame
+        Dataframe containing time series of discharge data to be plotted.
+    reference : str
+        Name of the reference data, must correspond to a column in the discharge
+        dataframe. Metrics are calculated between the reference column and each
+        of the other columns.
+    precipitation : pd.DataFrame, optional
+        Optional dataframe containing time series of precipitation data to be
+        plotted from the top of the hydrograph.
+    fname : str, Path
+        If specified, a copy of the plot will be saved to this path.
+    dpi : int, optional
+        DPI for the plot.
+    title : str, optional
+        Title of the hydrograph.
+    discharge_units : str, optional
+        Units for the discharge data.
+    precipitation_units : str, optional
+        Units for the precipitation data.
+
+    Returns
+    -------
+    ax : :obj:`matplotlib.axes.Axes`
+        Instance of a `~matplotlib.axes.Axes` object.
+    """
+    discharge_cols = discharge.columns.drop('reference')
+    y_obs = discharge[reference]
+    y_sim = discharge[discharge_cols]
+
+    fig, (ax, ax_tbl) = plt.subplots(
+        nrows=2,
+        ncols=1,
+        dpi=dpi,
+        figsize=(10, 10),
+        gridspec_kw={'height_ratios': [3, 1]},
```

Comment:
Doesn't this depend on the number of datasets you need to list in the table? I guess it's fine for now, but just not sure if it makes sense to hardcode it.

Reply:
It's all green to me, I cannot make up from your comment what it is that you mean.

Reply:
My comments always refer to the last line. I guess gh decided to print the preceding lines to give you some context, but apparently this is confusing rather than helpful.

If you have many rows in the table, the aspect ratio may no longer be suitable. I'm just trying to imagine edge cases where these hard-coded layout settings may fail.

Reply:
This implementation performs well with up to about 7. Any higher and you start to get some overlap with the x labels, and after 10 it overlaps with the data. We said it should work well with 3. Do you think we should support more than 7 in this PR?

Reply:
This can now be solved using the `figsize` argument if the height needs to be increased to make room for the table.